### ISSUE TITLE
Issues with using amp.initialize() and torch.FloatTensor 

### ISSUE THREAD

**Issue by mbufi** [NVIDIA/retinanet-examples#21]
Since the latest update and using the 19.04 container I cannot train models. Please see below:

`root@984a583da487:/retinanet-examples# retinanet train retinanet_rn101fpn_2.pth --backbone ResNet101FPN  --images coco/images/train2017/ --annotations coco/annotations/instances_train2017.json --val-images coco/images/val2017/ --val-annotations coco/annotations/instances_val2017.json
`
> Initializing model...
> Downloading: "https://download.pytorch.org/models/resnet101-5d3b4d8f.pth" to /root/.torch/models/resnet101-5d3b4d8f.pth
> 178728960it [00:31, 5759407.31it/s]
>      model: RetinaNet
>   backbone: ResNet101FPN
>    classes: 80, anchors: 9
> Selected optimization level O2:  FP16 training with FP32 batchnorm and FP32 master weights.
> 
> Defaults for this optimization level are:
> enabled                : True
> opt_level              : O2
> cast_model_type        : torch.float16
> patch_torch_functions  : False
> keep_batchnorm_fp32    : True
> master_weights         : True
> loss_scale             : dynamic
> Processing user overrides (additional kwargs that are not None)...
> After processing overrides, optimization options are:
> enabled                : True
> opt_level              : O2
> cast_model_type        : torch.float16
> patch_torch_functions  : False
> keep_batchnorm_fp32    : True
> master_weights         : True
> loss_scale             : 128.0
> Traceback (most recent call last):
>   File "/opt/conda/bin/retinanet", line 11, in <module>
>     load_entry_point('retinanet==0.1', 'console_scripts', 'retinanet')()
>   File "/opt/conda/lib/python3.6/site-packages/retinanet/main.py", line 181, in main
>     worker(0, args, 1, model, state)
>   File "/opt/conda/lib/python3.6/site-packages/retinanet/main.py", line 127, in worker
>     metrics_url=args.post_metrics, logdir=args.logdir, verbose=(rank == 0))
>   File "/opt/conda/lib/python3.6/site-packages/retinanet/train.py", line 34, in train
>     verbosity = is_master)
>   File "/opt/conda/lib/python3.6/site-packages/apex/amp/frontend.py", line 311, in initialize
>     return _initialize(models, optimizers, _amp_state.opt_properties)
>   File "/opt/conda/lib/python3.6/site-packages/apex/amp/_initialize.py", line 149, in _initialize
>     check_params_fp32(models)
>   File "/opt/conda/lib/python3.6/site-packages/apex/amp/_initialize.py", line 81, in check_params_fp32
>     name, param.type()))
>   File "/opt/conda/lib/python3.6/site-packages/apex/amp/_amp_state.py", line 31, in warn_or_err
>     raise RuntimeError(msg)
> RuntimeError: Found param backbones.ResNet101FPN.features.conv1.weight with type torch.FloatTensor, expected torch.cuda.FloatTensor.
> When using amp.initialize, you do not need to call .half() on your model
> before passing it, no matter what optimization level you choose.

**Comment by KodiaqQ** [comment]
Hi, @mbufi. Have you been solved this issue?

**Comment by florischabert** [comment]
Please try to update your driver to 418 for the pytorch:19.04 image. This error seems to be caused by the cuda compatibility layer with older versions of the driver.

**Comment by KodiaqQ** [comment]
Hmm. I have 418.56 driver version already and still have this error at 19.04 image.

**Comment by Maxfashko** [comment]
@florischabert i have the same problem. 
```
retinanet train \
  retinanet_rn152fpn.pth \
  --backbone ResNet152FPN \
  --batch 1 \
  --classes 80 \
  --iters 10000 \
  --lr 0.0005 \
  --val-iters 500 \
  --images /coco/train2017/ \
  --annotations /coco/annotations/instances_train2017.json \
  --val-images /coco/val2017/ \
  --val-annotations /coco/annotations/instances_val2017.json
```

> File "/opt/conda/bin/retinanet", line 11, in <module>
    load_entry_point('retinanet==0.1', 'console_scripts', 'retinanet')()
  File "/opt/conda/lib/python3.6/site-packages/retinanet-0.1-py3.6-linux-x86_64.egg/retinanet/main.py", line 181, in main
    worker(0, args, 1, model, state)
  File "/opt/conda/lib/python3.6/site-packages/retinanet-0.1-py3.6-linux-x86_64.egg/retinanet/main.py", line 127, in worker
    metrics_url=args.post_metrics, logdir=args.logdir, verbose=(rank == 0))
  File "/opt/conda/lib/python3.6/site-packages/retinanet-0.1-py3.6-linux-x86_64.egg/retinanet/train.py", line 153, in train
    mixed_precision=mixed_precision, is_master=is_master, world=world, use_dali=use_dali, verbose=False)
  File "/opt/conda/lib/python3.6/site-packages/retinanet-0.1-py3.6-linux-x86_64.egg/retinanet/infer.py", line 47, in infer
    verbosity = 0)
  File "/opt/conda/lib/python3.6/site-packages/apex/amp/frontend.py", line 311, in initialize
    return _initialize(models, optimizers, _amp_state.opt_properties)
  File "/opt/conda/lib/python3.6/site-packages/apex/amp/_initialize.py", line 149, in _initialize
    check_params_fp32(models)
  File "/opt/conda/lib/python3.6/site-packages/apex/amp/_initialize.py", line 81, in check_params_fp32
    name, param.type()))
  File "/opt/conda/lib/python3.6/site-packages/apex/amp/_amp_state.py", line 31, in warn_or_err
    raise RuntimeError(msg)
RuntimeError: Found param backbones.ResNet152FPN.features.conv1.weight with type torch.cuda.HalfTensor, expected torch.cuda.FloatTensor.
When using amp.initialize, you do not need to call .half() on your model
before passing it, no matter what optimization level you choose.

**Comment by florischabert** [comment]
This seems to come from the compat layer with some versions of the driver when using multiprocessing on the latest containers. You can try to `rm /etc/shinit` before running the `retinanet` command.
This should be fixed in some future version of the container.

**Comment by Maxfashko** [comment]
@florischabert it did not help

>[  120/90000] focal loss: 1.353, box loss: 0.464, 0.502s/2-batch (fw: 0.186s, bw: 0.313s), 4.0 im/s, lr: 0.0005
[  241/90000] focal loss: 1.287, box loss: 0.457, 0.498s/2-batch (fw: 0.181s, bw: 0.316s), 4.0 im/s, lr: 0.0005
[  366/90000] focal loss: 1.238, box loss: 0.460, 0.483s/2-batch (fw: 0.177s, bw: 0.306s), 4.1 im/s, lr: 0.0005
[  487/90000] focal loss: 1.267, box loss: 0.455, 0.497s/2-batch (fw: 0.182s, bw: 0.312s), 4.0 im/s, lr: 0.0005
Traceback (most recent call last):
  File "/opt/conda/bin/retinanet", line 11, in <module>
    load_entry_point('retinanet==0.1', 'console_scripts', 'retinanet')()
  File "/opt/conda/lib/python3.6/site-packages/retinanet-0.1-py3.6-linux-x86_64.egg/retinanet/main.py", line 181, in main
    worker(0, args, 1, model, state)
  File "/opt/conda/lib/python3.6/site-packages/retinanet-0.1-py3.6-linux-x86_64.egg/retinanet/main.py", line 127, in worker
    metrics_url=args.post_metrics, logdir=args.logdir, verbose=(rank == 0))
  File "/opt/conda/lib/python3.6/site-packages/retinanet-0.1-py3.6-linux-x86_64.egg/retinanet/train.py", line 153, in train
    mixed_precision=mixed_precision, is_master=is_master, world=world, use_dali=use_dali, verbose=False)
  File "/opt/conda/lib/python3.6/site-packages/retinanet-0.1-py3.6-linux-x86_64.egg/retinanet/infer.py", line 47, in infer
    verbosity = 0)
  File "/opt/conda/lib/python3.6/site-packages/apex/amp/frontend.py", line 311, in initialize
    return _initialize(models, optimizers, _amp_state.opt_properties)
  File "/opt/conda/lib/python3.6/site-packages/apex/amp/_initialize.py", line 149, in _initialize
    check_params_fp32(models)
  File "/opt/conda/lib/python3.6/site-packages/apex/amp/_initialize.py", line 81, in check_params_fp32
    name, param.type()))
  File "/opt/conda/lib/python3.6/site-packages/apex/amp/_amp_state.py", line 31, in warn_or_err
    raise RuntimeError(msg)
RuntimeError: Found param backbones.ResNet50FPN.features.conv1.weight with type torch.cuda.HalfTensor, expected torch.cuda.FloatTensor.
When using amp.initialize, you do not need to call .half() on your model
before passing it, no matter what optimization level you choose.

**Comment by pkashinkunti** [comment]
So, it appears that there are two separate issues going on here:

1.) Issues stemming from the compat layer that cause training to fail right away (e.g. `RuntimeError: Found param backbones.ResNet101FPN.features.conv1.weight with type **torch.FloatTensor, expected torch.cuda.FloatTensor.**`)

2.) Issues occuring during the first validation inference like @Maxfashko is referencing (e.g. `RuntimeError: Found param backbones.ResNet50FPN.features.conv1.weight with type **torch.cuda.HalfTensor, expected torch.cuda.FloatTensor.**`)

From our testing, 1.) should be resolved by updating the driver to 418 or by using the `rm /etc/shinit` workaround mentioned above.

2.) looks like a bug in the inference logic: This should be resolved with #29 which has just been merged. 

I am going to close this issue: @mbufi, if you are still seeing an issue like 1.) after updating the driver or using the workaround, please reopen and provide some additional details on your specific setup. @Maxfashko, please try with the latest updates and open a separate issue if that does not resolve.
